protocol/contracts/oud/OudRedeemer.sol

```diff
@@ -0,0 +1,145 @@
+  function treasuryPriceIndex() external view returns (uint256) {
+    return tpi;
+  }
+
+  // @dev Allows the tpi to be adjusted.
+  // @param 'value' is the new TPI
+  function setTreasuryPriceIndex(uint256 value) external override onlyOwner {
+    tpi = value;
+  }
+
+  function setStableCoin(address newStable) external override onlyOwner {
+    if (newStable == address(0)) revert AddressZero();
+    stable = IERC20(newStable);
+  }
+
+  // @dev Decimal precision for 'tpi', 9880 == $0.988, precision = 4
+  function TPI_PRECISION() external pure returns (uint256) {
+    return tpiDecimals;
+  }
+
+  /**
+   * @notice Provides a quote for 'oudAmount' with relevant token addresses:
+   *  'oudAmount' provided, 'stable' token required, 'templeAmount' that will be minted to caller
+   * @param oudAmount amount of Oud that will be redeemed
+   * @return _oudAmount amount of Oud that wil be redeemed
+   * @return _oudTokenAddress address of the Oud token
+   * @return _stableAmount amount of Stable token required at tpi
+   * @return _stableAddress address of the Stable token in use
+   * @return _templeAmount amount of Temple that will be minted
+   * @return _templeTokenAddress address of the Temple token
+   */
+  function getQuote(
+    uint256 oudAmount
+  )
+    external
+    view
+    returns (
+      uint256 _oudAmount,
+      address _oudTokenAddress,
+      uint256 _stableAmount,
+      address _stableAddress,
+      uint256 _templeAmount,
+      address _templeTokenAddress
+    )
+  {
+    return (
+      oudAmount,
+      oudToken,
+      _getStableAmount(oudAmount),
+      address(stable),
+      oudAmount,
+      templeToken
+    );
+  }
+
+  // @notice Redeems Oud and Stable token (@ tpi) for Temple
+  // @dev requires allowances for Oud and Stable Token
+  // @param 'oudAmount' The amount of Oud to be redeemed for Temple
+  // @returns 'templeAmount' The amount of Temple that has been minted to sender.
+  function redeem(uint256 oudAmount) external returns (uint256 templeAmount) {
+    if (oudAmount == 0) revert RedeemAmountZero();
+    address account = msg.sender;
+    uint256 _stableAmount;
+    _stableAmount = _getStableAmount(oudAmount);
+    IOudToken(oudToken).burn(account, oudAmount);
+    SafeERC20.safeTransferFrom(stable, account, address(this), _stableAmount);
+    SafeERC20.safeIncreaseAllowance(stable, depositStableTo, _stableAmount);
+    Oud__ITempleLineOfCredit(depositStableTo).depositReserve(
+      stable,
+      address(this),
+      _stableAmount
+    );
+
+    // @dev Conversion is 1 OUD + (Stable*TPI) = 1 Temple thus Temple Amount == Oud Amount.
+    Oud__ITempleERC20Token(templeToken).mint(account, oudAmount);
+    emit OudRedeemed(account, oudAmount, tpi, oudAmount);
+    return oudAmount;
+  }
+
+  function _getStableAmount(
+    uint256 oudAmount
+  ) internal view returns (uint256 stableAmount) {
+    return (oudAmount * tpi) / (10 ** tpiDecimals);
+  }
+
+  event OudRedeemed(
```

Comment:
Best to put this up the top of the contract along side the errors

Reply:
yup, sorry, was looking at vault.sol at the time, assumed you look at one contract, and they all follow the same standard.

That being said, most have it just below the variables, above the constructor?